### PR TITLE
add stake lockup

### LINKED ISSUE #5781: stake lockup
#### Problem
 grants placed in stake accounts will require a lockup, some term during which the lamports are unavailable for withdrawal

#### Proposed Solution
* add lockup, a target slot height at which the stake lamports are available for withdrawal
* move stake state through a lockup state that looks like "uninitialized" to be able to lockup future grants without naming a delegate

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem
 grants placed in stake accounts will require a lockup, some term during which the lamports are unavailable for withdrawal

 #### Summary of Changes
 add lockup stuff
 move creation through lockup
 some nits

Fixes #5781

### CODE REVIEW

File: cli/src/wallet.rs [hunk 1]
```diff
@@ -895,7 +895,9 @@ fn process_show_stake_account(
             Ok("".to_string())
         }
         Ok(StakeState::RewardsPool) => Ok("Stake account is a rewards pool".to_string()),
-        Ok(StakeState::Uninitialized) => Ok("Stake account is uninitialized".to_string()),
+        Ok(StakeState::Uninitialized) | Ok(StakeState::Lockup(_)) => {
+            Ok("Stake account is uninitialized".to_string())
```
Comment (aeyakovenko) [inline cli/src/wallet.rs]: probably want a different error for each branch just for debugging

File: programs/stake_api/src/stake_state.rs [hunk 2]
```diff
@@ -326,13 +340,14 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
             return Err(InstructionError::InsufficientFunds);
         }
 
-        if let StakeState::Uninitialized = self.state()? {
+        if let StakeState::Lockup(lockup) = self.state()? {
```
Comment (aeyakovenko) [inline programs/stake_api/src/stake_state.rs]: stakes should be delegatable if they are locked up, just claiming rewards is locked

File: cli/src/wallet.rs (hunk 1 above)
Comment (rob-solana) [reply cli/src/wallet.rs]: I have an issue specifically for errors in stake_api

File: programs/stake_api/src/stake_state.rs (hunk 2 above)
Comment (rob-solana) [reply programs/stake_api/src/stake_state.rs]: yes, that's what this says

Comment (rob-solana) [reply programs/stake_api/src/stake_state.rs]: StakeState::Uninitialized is fleeting, we set the stake to lockup at creation time.  the owner of the account can later delegate (and I'm going to also add re-delegation)

**Comment by aeyakovenko** [comment]
just so I grok this

* stake gets created with a lockup
* rewards can be claimed
* withdraw is disabled until lockup is reached

Are claimed rewards automatically staked?

**Comment by rob-solana** [comment]
> just so I grok this
> 
> * stake gets created with a lockup
> * rewards can be claimed
> * withdraw is disabled until lockup is reached
> 
> Are claimed rewards automatically staked?



> just so I grok this
> 
> * stake gets created with a lockup

yep, and can be delegated by the owner

> * rewards can be claimed

yep

> * withdraw is disabled until lockup is reached

yep

> Are claimed rewards automatically staked?

claimed rewards are not automatically staked.  I haven't figured out a good way to do that without a table of when the rewards were generated and then activated.  imagine a stake account that's been generating rewards for a really long time without claims.  suddenly there's a big balance jump, and possibly a big stake jump.

**Comment by rob-solana** [comment]
what I could do is support creation of another stake account with the same lockup using rewards proceeds (or any amount of lamports in excess of stake)

**Pull request merged**